Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx

```diff
@@ -230,14 +257,297 @@ ContourExtractor2DImageFilter<TInputImage>::GenerateData()
+    RegionConstIterator inIt{ this->GetInput(), inputRegion };
+    for (inIt.GoToBegin(); !inIt.IsAtEnd(); ++inIt)
+    {
+      allLabels.push_back(inIt.Get());
+    }
+    std::sort(allLabels.begin(), allLabels.end());
+    const typename std::vector<InputPixelType>::const_iterator last{ std::unique(allLabels.begin(), allLabels.end()) };
+    allLabels.erase(last, allLabels.end());
+  }
+
+  // Make room so that largerImage and its largerRegion can have a one-pixel wide border on all sides of the input
+  // region.
+  const InputImagePointer largerImage{ InputImageType::New() };
+  const IndexType         largerIndex{ inputRegion.GetIndex()[0] - 1, inputRegion.GetIndex()[1] - 1 };
+  const SizeType          largerSize{ inputRegion.GetSize()[0] + 2, inputRegion.GetSize()[1] + 2 };
+  const InputRegionType   largerRegion{ largerIndex, largerSize };
+  largerImage->SetRegions(largerRegion);
+  largerImage->Allocate();
+  {
+    RegionConstIterator inputIt{ this->GetInput(), inputRegion };
+    RegionIterator      largerIt{ largerImage, inputRegion };
+    for (inputIt.GoToBegin(), largerIt.GoToBegin(); !inputIt.IsAtEnd(); ++inputIt, ++largerIt)
+    {
+      largerIt.Set(inputIt.Get());
+    }
+  }
+
+  // Set up an iterator to "march the squares" across the image.
+  // We associate each 2px-by-2px square with the pixel in the top left of
+  // that square. We then iterate across the image, examining these 2x2 squares
+  // and building the contour. By iterating the top-left pixel of our
+  // "current square" across every pixel in the image except those on the
+  // bottom row and rightmost column, we have visited every valid square in the
+  // image.
+
+  constexpr RadiusType      radius = { { 1, 1 } };
+  constexpr InputOffsetType none = { { 0, 0 } };
+  constexpr InputOffsetType right = { { 1, 0 } };
+  constexpr InputOffsetType down = { { 0, 1 } };
+  constexpr InputOffsetType diag = { { 1, 1 } };
+
+  // Save m_ContourValue because we are going to overwrite with a
+  // value half way between false (0) and true (1).
+  const InputRealType save_ContourValue{ this->GetContourValue() };
+  this->SetContourValue(0.5);
+
+  for (InputPixelType label : allLabels)
+  {
+    // Make sure the structures for containing, looking up, and numbering the
+    // growing contours are empty and ready.
+    m_Contours.clear();
+    m_ContourStarts.clear();
+    m_ContourEnds.clear();
+    m_NumberOfContoursCreated = 0;
+
+    // Find a bounding box for this label so that we can skip much of the region
+    const IndexType left_top{ inputRegion.GetIndex() };
+    const IndexType right_bot{ inputRegion.GetIndex()[0] + static_cast<IndexValueType>(inputRegion.GetSize()[0]) - 1,
+                               inputRegion.GetIndex()[1] + static_cast<IndexValueType>(inputRegion.GetSize()[1]) - 1 };
+    IndexType       max = left_top;
+    IndexType       min = right_bot;
+    {
+      RegionConstIterator inputIt{ this->GetInput(), inputRegion };
```

Comment:
@thewtex In this particular case, both the pixel value `inputIt.Get()` and its location `inputIt.GetIndex()` are retrieved, for each iteration. I'm afraid that isn't yet supported by the ITK5 ranges. We implemented ranges of pixel values and [ranges of indices](https://itk.org/Doxygen/html/classitk_1_1IndexRange.html), but we did not yet support ranges of pixel+index pairs.

Reply:
Using `ImageRegionRange` instead.

Reply:
Using `ImageRegionRange` and `ImageRegionIndexRange`, it could look like this:

      auto inputImageRegionRangeIterator = inputImageRegionRange.cbegin();
      for (const auto index : ImageRegionIndexRange<2>(inputRegion))
      {
        if (*inputImageRegionRangeIterator == label)
        {
          max[0] = std::max(max[0], index[0]);
          max[1] = std::max(max[1], index[1]);
          min[0] = std::min(min[0], index[0]);
          min[1] = std::min(min[1], index[1]);
        }
        ++inputImageRegionRangeIterator;
      }

But I'm not 100% sure if it is really superior to the original code 🤔  Because it uses two ranges (a pixel value range and an index range), instead of one.

Reply:
I'll leave it as is unless anyone has any strong opinions here.